rs/execution_environment/fuzz/src/lib.rs

```diff
@@ -59,3 +79,145 @@ pub fn fuzzer_main() {
+                Sysno::futex,
+                Sysno::close,
+            ]);
+            loop {
+                // This code employs a manual heuristic to determine which process PID to attach to,
+                // specifically targeting the one executing the Wasm code.
+                //
+                // The child process spawns a total of 14 threads. Since PIDs are monotonically increasing
+                // and the tid are stored in a BTreeSet, they are ordered based on their creation sequence.
+                //
+                // By tracing all PIDs and analyzing the associated syscalls, we observe that the critical
+                // threads to attach to are typically among the last few, specifically [n-2] and [n-1].
+                //
+                // NOTE: If the code design changes in the future, this heuristic will need to be revisited
+                // and updated accordingly.
+                let mut children = get_children(child.into());
+                for _ in 0..1 {
+                    children.pop_last();
+                }
+                let child = children.last().unwrap();
+                trace(name, Pid::from_raw(*child), &allowed_syscalls);
+            }
+        }
+        Err(err) => {
+            panic!("{} fork() failed: {}", name, err);
+        }
+    }
+}
+
+#[cfg(not(target_os = "linux"))]
+fn syscall_monitor<F>(_name: &str, sandbox: F)
+where
+    F: Fn(),
+{
+    sandbox();
+}
+
+#[cfg(target_os = "linux")]
+fn trace(name: &str, child: Pid, allowed_syscalls: &BTreeSet<Sysno>) {
+    if let Err(err) = ptrace::attach(child) {
+        println!(
+            "ptrace: failed to attach process {}::{}: {}",
+            name, child, err
+        );
+        return;
+    }
+
+    while let Ok(result) = waitpid(child, None) {
+        match result {
+            WaitStatus::Stopped(..) => {
+                if let Err(err) = ptrace::setoptions(child, Options::all()) {
+                    panic!(
+                        "ptrace: failed to setoptions process {}::{}: {}",
+                        name, child, err
+                    );
+                }
+
+                if let Ok(regs) = ptrace::getregs(child) {
+                    let sysno = Sysno::from(regs.orig_rax as u32);
+                    if !allowed_syscalls.contains(&sysno) {
+                        panic!("Syscall not present: {:?} {}::{}", sysno, name, child,);
+                    }
+                }
+
+                if let Err(err) = ptrace::syscall(child, None) {
+                    panic!(
+                        "ptrace: failed to continue to next syscall {}::{}: {}",
+                        name, child, err
+                    );
+                }
+            }
+            WaitStatus::PtraceSyscall(_) => {
+                if let Ok(regs) = ptrace::getregs(child) {
+                    let sysno = Sysno::from(regs.orig_rax as u32);
+                    if !allowed_syscalls.contains(&sysno) {
```

Comment:
We should likely track entries to and exits from system calls. On syscall exit, RAX contains the return value...

Reply:
Done and with `PTRACE_O_TRACESYSGOOD` set via Options:all, it should be easy as flipping a boolean in the while loop. However, I did notice the `rax` being the same on both entry and exit so maybe I'm missing something.

Reply:
Sorry, `orig_rax` always contains the syscall number, `rax` contains the return value on exit.